### PR TITLE
feat: allow using a local RSA key for machine keys

### LINKED ISSUE #7554: Locally generate private key for service user
### Preflight Checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions
- [X] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

Private keys generally should not be sent over the network. I would like to be able to create a service user with a private key that has been generated locally.

### Describe your ideal solution

Add an optional parameter to the request body of the [add machine key](https://zitadel.com/docs/apis/resources/mgmt/management-service-add-machine-key) which accepts an RSA public key. The JSON returned by the endpoint has the same fields as before but is missing the `key` field.

```json
{
    "type":"serviceaccount",
    "keyId":"100509901696068329",
    "userId": "100507859606888466"
}
```

### Version

v2.46.0

### Environment

Self-hosted

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by aDogCalledSpot**
Closes #7554 

Adds the option of using a locally generated RSA key for a machine key instead of having it generated by the server.
This prevents sending private keys over the network which is generally advised against.

A user can use the optional "pubkey" field of the request to send a public key. If a public key is specified, then no private key will be returned. If no public key is specified the behaviour is as before.

### Definition of Ready

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding ~~user story~~ issue
- [ ] Acceptance criteria are met
- [ ] ~~All open todos and follow ups are defined in a new ticket and justified~~
- [ ] ~~Deviations from the acceptance criteria and design are agreed with the PO and documented~~
- [x] No debug or dead code
- [x] My code has no repetitions
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

**Comment by aDogCalledSpot** [comment]
@livio-a

File: proto/zitadel/management.proto [hunk 1]
```diff
@@ -8504,6 +8504,7 @@ message AddMachineKeyRequest {
             description: "The date the key will expire and no logins will be possible";
         }
     ];
+    bytes pubkey = 4;
```
Comment (livio-a) [inline proto/zitadel/management.proto:8507]:
```suggestion
    bytes public_key = 4 [
        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
           example: "\"LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCk1...\"";
           description: "Optionally provide a public key of your own generated RSA private key.";
        }
    ];
```

**Review by livio-a — CHANGES_REQUESTED** [review]
Hey @aDogCalledSpot 

That already works perfectly 😃  
I'd change the proto slightly and think we should also check if the bytes provided can be parsed into an RSA public key.
This could be handled in:
https://github.com/zitadel/zitadel/blob/dff5984f7db1c5326cffc1ab32baf9cde882dcf3/internal/command/user_machine_key.go#L77-L83
with something like:
```go
if _, err := crypto.BytesToPublicKey(key.PublicKey); err != nil {
	return zerrors.ThrowInvalidArgument(nil, "COMMAND-5F3h1", "Errors.User.Machine.Key.Invalid")
}
```
And the translations for the new error `Errors.User.Machine.Key.Invalid` would then need to be added in https://github.com/zitadel/zitadel/blob/dff5984f7db1c5326cffc1ab32baf9cde882dcf3/internal/static/i18n/en.yaml#L108-L111 and the corresponding files.

**Comment by aDogCalledSpot** [comment]
> return zerrors.ThrowInvalidArgument(nil, "COMMAND-5F3h1", "Errors.User.Machine.Key.Invalid")

Where does the ID `5F3h1` come from? Does it need to be documented anywhere?

**Comment by livio-a** [comment]
> > return zerrors.ThrowInvalidArgument(nil, "COMMAND-5F3h1", "Errors.User.Machine.Key.Invalid")
> 
> Where does the ID `5F3h1` come from? Does it need to be documented anywhere?

That's randomly generated. You can just use it in that case.

**Comment by aDogCalledSpot** [comment]
@livio-a can you re-review?

**Comment by aDogCalledSpot** [comment]
Can we re-run the CI?

**Comment by aDogCalledSpot** [comment]
> Can we re-run the CI?

bump

**Comment by livio-a** [comment]
> > Can we re-run the CI?
> 
> bump

Sorry was on vacation. Just had a look at it. Can you also provide the translations for the other languages, please.

**Comment by livio-a** [comment]
> > Can you also provide the translations for the other languages, please
> 
> I can provide it for German but otherwise all I could do is put it in google translate/deepL. Is that sufficient?

That's totally ok. I would need to do the same 😄

**Review by livio-a — APPROVED** [review]

**Comment by livio-a** [comment]
@aDogCalledSpot thanks for this contribution. Really appreciated!
It will be released within this week 😃 
If you like to get a gift from us, please write an email to [hi@zitadel.com](mailto:hi@zitadel.com)

**Pull request merged**